V11的合并转发再次无法发送

最新版的`onebots`和`icqqjs/icqq` 又出现了合并转发一直超时的问题，我这边用的时候好像 action 匹配不到就会一直超时，应该是 V11 没有像 V12 那样匹配不到 action 就抛出 404 异常的兜底。

> 请提供一下log日志和代码截图

```
[2024-03-25T15:43:03.763] [INFO] [OneBots:icqq-V11(**********)] -  收到ws消息： {"action":"send_private_msg","params":{"user_id":*******,"message":[{"type":"node","data":{"user_id":**********,"nickname":"派蒙","content":[{"type":"text","data":{"text":"测试内容1"}}]}},{"type":"node","data":{"user_id":**********,"nickname":"派蒙","content":[{"type":"text","data":{"text":"测试内容2"}}]}},{"type":"node","data":{"user_id":**********,"nickname":"派蒙","content":[{"type":"text","data":{"text":"测试内容3"}}]}}]},"echo":"lu6n54mjP9t86b"}
[2024-03-25T15:43:03.767] [ERROR] [OneBots:icqq-V11(**********)] - run send_private_msg with args:3 failed: TypeError: Cannot read properties of undefined (reading 'user_id')
    at IcqqAdapter.processMessages (/app/node_modules/onebots/lib/adapters/icqq/utils.js:14:35)
    at IcqqAdapter.sendPrivateMessage (/app/node_modules/onebots/lib/adapters/icqq/index.js:142:101)
    at IcqqAdapter.call (/app/node_modules/onebots/lib/adapters/icqq/index.js:192:36)
    at V11.sendPrivateMsg (/app/node_modules/onebots/lib/service/V11/action/friend.js:14:29)
    at V11.apply (/app/node_modules/onebots/lib/service/V11/index.js:594:53)
    at WebSocket.<anonymous> (/app/node_modules/onebots/lib/service/V11/index.js:420:38)
    at WebSocket.emit (node:events:517:28)
    at Receiver.receiverOnMessage (/app/node_modules/ws/lib/websocket.js:1209:20)
    at Receiver.emit (node:events:517:28)
    at /app/node_modules/ws/lib/receiver.js:608:16
```

顺便说下 Web 的 Console 好像不能选中复制日志？

--- a/client/App.vue
+++ b/client/App.vue
@@ -96,7 +96,7 @@
                 </el-descriptions-item>
                 {{ systemInfo }}
             </el-descriptions>
-            <pre @click="input?.focus?.()">{{ logs }}</pre>
+            <pre>{{ logs }}</pre>
             <div class="input-wrapper">
                 <span class="text">{{ inputData }}</span>
                 <input ref="input" v-model="inputData" @keyup.enter="submitInput" />
@@ -109,6 +109,7 @@
 import { nextTick, onMounted, onUnmounted, ref } from "vue";
 import { AdapterInfo, SystemInfo } from "./types";
 import { formatSize, formatTime } from "./utils";
+
 const input = ref();
 const ws = ref<WebSocket>();
 const config = ref<string>("");
@@ -226,7 +227,8 @@ const init = () => {
     const port =
         localStorage.getItem("OneBots:serverPort") ||
         prompt("请输入服务端监听的端口号", location.port);
-    const wsUrl = `${location.protocol.replace("http", "ws")}//${location.hostname}:${port}`;
+    const wsProtocol = location.protocol === "https:" ? "wss:" : "ws:";
+    const wsUrl = `${wsProtocol}//${location.hostname}:${port}`;
     ws.value = new WebSocket(wsUrl);
     ws.value.onerror = e => {
         console.error("连接出错", e);

--- a/src/adapters/icqq/index.ts
+++ b/src/adapters/icqq/index.ts
@@ -204,7 +204,7 @@ export default class IcqqAdapter extends Adapter<"icqq", Sendable> {
         const oneBot = this.getOneBot<Client>(uin);
         if (!oneBot) throw new Error("No one");
         let { message, ...result } = await oneBot.internal.getMsg.call(oneBot.internal, message_id);
-        const segments = this.toSegment(version, message);
+        const segments = version === "V11" ? message : this.toSegment(version, message);
         return {
             ...result,
             message: segments,

--- a/src/adapters/icqq/index.ts
+++ b/src/adapters/icqq/index.ts
@@ -34,7 +34,7 @@ export default class IcqqAdapter extends Adapter<"icqq", Sendable> {
         const oneBot = this.getOneBot<Client>(uin);
         if (!oneBot) throw new Error("No one");
         if (!oneBot.internal[name]) throw new Error(`internal no api ${name}`);
-        return oneBot[version][name](...args);
+        return oneBot.internal[name](...args);
     }
 
     createOneBot(uin: string, protocol: IcqqConfig, versions: OneBot.Config[]): OneBot {

--- a/src/adapters/icqq/utils.ts
+++ b/src/adapters/icqq/utils.ts
@@ -15,16 +15,17 @@ export async function processMessages(
         const { type, data, ...other } = item;
         switch (type) {
             case "node": {
+                const node = data || { ...other };
                 result.push({
                     type,
-                    ...(data || other),
-                    user_id: data.user_id,
+                    ...node,
+                    user_id: node.user_id,
                     message: await processMessages.call(
                         this,
                         uin,
-                        data.user_id,
+                        node.user_id,
                         "private",
-                        data.content || [],
+                        node.content || []
                     ),
                 });
                 break;

--- a/src/server/app.ts
+++ b/src/server/app.ts
@@ -1,17 +1,17 @@
 import Koa from "koa";
 import * as os from "os";
 import "reflect-metadata";
-import { copyFileSync, existsSync, mkdirSync, writeFileSync } from "fs";
-import { Logger, getLogger, configure } from "log4js";
+import * as fs from "fs";
+import { copyFileSync, existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
+import { configure, getLogger, Logger } from "log4js";
 import { createServer, Server } from "http";
 import koaStatic from "koa-static";
 import yaml from "js-yaml";
 import KoaBodyParser from "koa-bodyparser";
 import basicAuth from "koa-basic-auth";
 
-import { deepMerge, deepClone, Class, readLine } from "@/utils";
+import { Class, deepClone, deepMerge, readLine } from "@/utils";
 import { Router, WsServer } from "./router";
-import { readFileSync } from "fs";
 import { V11 } from "@/service/V11";
 import { V12 } from "@/service/V12";
 import { LogLevel } from "@/types";
@@ -19,8 +19,8 @@ import * as path from "path";
 import { Adapter } from "@/adapter";
 import { ChildProcess } from "child_process";
 import process from "process";
-import * as fs from "fs";
 import { Dict } from "@zhinjs/shared";
+
 export interface KoaOptions {
     env?: string;
     keys?: string[];
@@ -216,7 +216,22 @@ export class App extends Koa {
                 }
                 switch (payload.action) {
                     case "system.input":
-                        return process.stdin.write(`${payload.data}\n`);
+                        // 将流的模式切换到“流动模式”
+                        process.stdin.resume();
+
+                        // 使用以下函数来模拟输入数据
+                    function simulateInput(data: Buffer) {
+                        process.nextTick(() => {
+                            process.stdin.emit("data", data);
+                        });
+                    }
+
+                        simulateInput(Buffer.from(payload.data + "\n", "utf8"));
+                        // 模拟结束
+                        process.nextTick(() => {
+                            process.stdin.emit("end");
+                        });
+                        return true;
                     case "system.saveConfig":
                         return fs.writeFileSync(App.configPath, payload.data, "utf8");
                     case "system.reload":

--- a/src/service/V11/index.ts
+++ b/src/service/V11/index.ts
@@ -1,14 +1,13 @@
 import { Config } from "./config";
 import { Action } from "./action";
-import { OneBot, OneBotStatus } from "@/onebot";
+import { BOOLS, NotFoundError, OneBot, OneBotStatus } from "@/onebot";
 import { Logger } from "log4js";
 import crypto from "crypto";
 import { WebSocket, WebSocketServer } from "ws";
 import { Dispose } from "@/types";
 import { Context } from "koa";
 import { URL } from "url";
-import { toBool, toHump, toLine, randomInt } from "@/utils";
-import { BOOLS, NotFoundError } from "@/onebot";
+import { randomInt, toBool, toHump, toLine } from "@/utils";
 import http from "http";
 import https from "https";
 import { join } from "path";
@@ -621,7 +620,7 @@ export class V11 extends Service<"V11"> implements OneBot.Base {
                     ret = await this.action[method].apply(this, args);
                 } catch (e) {
                     this.logger.error(`run ${action} with args:${args.length} failed:`, e);
-                    return JSON.stringify(V11.error(e.message));
+                    return JSON.stringify(V11.error(e.message, echo));
                 }
                 if (ret instanceof Promise) {
                     if (is_async) {
@@ -640,7 +639,7 @@ export class V11 extends Service<"V11"> implements OneBot.Base {
                 result.echo = echo;
             }
             return JSON.stringify(result, (_, v) => (typeof v === "bigint" ? v.toString() : v));
-        }
+        } else throw new NotFoundError();
     }
 
     /**
@@ -667,24 +666,30 @@ export namespace V11 {
         status: "ok" | "async" | "error";
         data: T;
         error: string;
+        msg?: string;
+        wording?: string;
         echo?: string;
     }
 
-    export function ok<T extends any>(data: T, retcode = 0, pending?: boolean): Result<T> {
+    export function ok<T extends any>(data: T, retcode = 0, pending?: boolean, echo?: string): Result<T> {
         return {
-            retcode,
+            retcode: pending ? 1 : retcode,
             status: pending ? "async" : "ok",
             data,
             error: null,
+            echo
         };
     }
 
-    export function error(error: string, retcode = 1): Result<any> {
+    export function error(error: string, echo?: string, retcode = 1500, wording?: string): Result<any> {
         return {
             retcode,
             status: "error",
             data: null,
             error,
+            msg: error,
+            wording,
+            echo
         };
     }
 

